### PR TITLE
Code clean-up for local kwargs in MCMCSamples.plot_2d

### PR DESCRIPTION
**Pull request by williamjameshandley**
# Description

This follows from PR #10, defining a local dictionary of keyword arguments for easier iteration.

@lukashergt -- what do you think of  the compactness/clarity of this code?

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code is PEP8 compliant (`flake8 anesthetic tests`)
- [x] My code contains compliant docstrings (`pydocstyle --convention=numpy anesthetic`)
- [x] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by lukashergt** [comment]
Love it. Gosh, you are good at this!

**Comment by williamjameshandley** [comment]
If you approve, then you can merge in via the GitHub interface if you wish

File: anesthetic/samples.py [hunk 1]
```diff
@@ -206,33 +206,19 @@ def plot_2d(self, axes, *args, **kwargs):
...
-        upper_kwargs: dict, optional
-            kwargs only for the upper 2D plots. This is useful when there
-            is a conflict of kwargs for different types of plots.
-            Note that any kwargs directly passed to plot_2d will overwrite any
-            kwarg with the same key passed to upper_kwargs.
+        diagonal_kwargs, lower_kwargs, upper_kwargs: dict, optional
+            kwargs for the diagonal (1D)/lower or upper (2D) plots. This is
+            useful when there is a conflict of kwargs for different types of
+            plots.  Note that any kwargs directly passed to plot_2d will
+            overwrite any kwarg with the same key passed to diagonal_kwargs.
```
Comment (lukashergt) [inline anesthetic/samples.py]: Nicely reduces the repetition. Last reference to `diagonal_kwargs` should be changed to `sub-kwargs` or a mention of all of them, though.

File: anesthetic/samples.py [hunk 2]
```diff
@@ -206,33 +206,19 @@ def plot_2d(self, axes, *args, **kwargs):
         types: dict, optional
             What type (or types) of plots to produce. Takes the keys 'diagonal'
             for the 1D plots and 'lower' and 'upper' for the 2D plots.
-            The options for 'diagonal are either:
+            The options for 'diagonal are:
```
Comment (lukashergt) [inline anesthetic/samples.py]: If you dislike the 'either', should it be removed for the parameter `axes` as well?

File: anesthetic/samples.py [hunk 3]
```diff
@@ -282,18 +267,15 @@ def plot_2d(self, axes, *args, **kwargs):
...
                     elif ax._upper:
-                        plot_type = types['upper']
-                        kwargs = upper_kwargs
+                        pos = 'upper'
                     else:
-                        plot_type = types['lower']
-                        kwargs = lower_kwargs
-                    self.plot(ax_, x, y, plot_type=plot_type, *args, **kwargs)
+                        pos = 'lower'
+                    ax_ = ax.twin if pos == 'diagonal' else ax
```
Comment (lukashergt) [inline anesthetic/samples.py]: I find `ax_ = ax.twin` shorter than this line. Also one conditional check less.

File: anesthetic/samples.py [hunk 4]
```diff
@@ -264,12 +250,11 @@ def plot_2d(self, axes, *args, **kwargs):
...
                 types = {'lower': types[0], 'upper': types[-1]}
-        diagonal_kwargs = kwargs.pop('diagonal_kwargs', {})
-        lower_kwargs = kwargs.pop('lower_kwargs', {})
-        upper_kwargs = kwargs.pop('upper_kwargs', {})
-        diagonal_kwargs.update(kwargs)
-        lower_kwargs.update(kwargs)
-        upper_kwargs.update(kwargs)
+
+        local_kwargs = {pos: kwargs.pop('%s_kwargs' % pos, {})
+                        for pos in ['lower', 'upper', 'diagonal']}
```
Comment (lukashergt) [inline anesthetic/samples.py]: lol, and we've got our dict in dict after all... at least we are sparing the user of having to deal with it...

File: anesthetic/samples.py (hunk 3 above)
Comment (williamjameshandley) [reply anesthetic/samples.py]: Agreed. One-line conditionals are not very pythonic

File: anesthetic/samples.py (hunk 2 above)
Comment (williamjameshandley) [reply anesthetic/samples.py]: Indeed. I've also cleaned up some of the other lists in the documentation of this file

File: anesthetic/samples.py (hunk 1 above)
Comment (williamjameshandley) [reply anesthetic/samples.py]: Good spot

**Comment by williamjameshandley** [comment]
So the tests were checking code coverage, which I put in to encourage increasing the number of tests.

I found a few errors in our implementation (which proves that this check is working). I also found some bugs that need fixing (will come in another pull request).

File: anesthetic/samples.py [hunk 5]
```diff
@@ -24,7 +24,7 @@ class MCMCSamples(WeightedDataFrame):
     Example plotting commands include
         - ``mcmc.plot_1d(['paramA', 'paramB'])``
         - ``mcmc.plot_2d(['paramA', 'paramB'])``
-        - ``mcmc.plot_2d(['paramA', 'paramB'], ['paramC', 'paramD'])``
+        - ``mcmc.plot_2d([['paramA', 'paramB'], ['paramC', 'paramD'])]``
```
Comment (lukashergt) [inline anesthetic/samples.py]: Closing brackets wrong way round: should be ]) but is )]

Comment (williamjameshandley) [reply anesthetic/samples.py]: In theory there are test suites that can test docstring code, but for now we'll need to rely on eagle-eyed reviewers like yourself

**Pull request merged**